scripts/src/filter_raw_reads.R

```diff
@@ -38,47 +73,77 @@ suppressPackageStartupMessages(library(sets))
+  # Validation: Check that id_cols exist in the sample meta ----
+  if(!validate_columns_exist(id_cols, sample_meta)) {
+    stop('One or more id_cols is NOT present in the sample meta.')
+  }
+  
+  # Validation: Check that sequencing_index_cols uniquely identify every rows of sample meta ----
+  if(!validate_unique_samples(sequencing_index_cols, sample_meta)) {
+    print('There may be multiple entries in the sample meta that have the same combination of sequencing index columns.')
+    stop('The specified sequencing index columns do NOT uniquely identify every PCR well.')
+  }
   
-  # index purity for QC table
+  # Filtering by sequencing columns ----
+  # Filter raw counts using the sequencing columns. 
+  # Also create "mapped" column to identify reads that mapped to all known PRISM sequences.
+  print("Filtering by sequencing columns ...")
+  unique_sequencing_index_vals= sample_meta %>% dplyr::distinct(pick(all_of(sequencing_index_cols)))
+  index_filtered= raw_counts %>% dplyr::semi_join(unique_sequencing_index_vals, by= sequencing_index_cols) %>%
+    dplyr::mutate(mapped= forward_read_cl_barcode %in% c(cell_line_meta$Sequence, CB_meta$Sequence))
+  
+  # Calculate index purity for QC table.
   index_purity= sum(index_filtered$n)/ sum(raw_counts$n)
   
-  # Mapping barcodes with valid indices ----
+  # Split off unmapped reads ----
+  # Unmapped reads are defined as having valid indices but do not map to barcodes in PRISM.
+  # Also sorted reads in descending order by read count.
   print('Splitting off unmapped reads ...')
-  # unmapped reads are defined as having valid indices but do not map to barcodes in PRISM
   unmapped_reads= index_filtered %>% dplyr::filter(mapped==F) %>% dplyr::select(-mapped) %>% 
     dplyr::arrange(dplyr::desc(n))
   
-  # make template of expected reads
+  # Creating a template of all expected reads in the run ----
+  # Use all 4 meta data files to create a "template" dataframe where
+  # every row is a cell line that is expected in a PCR well. 
   print('Creating template of expected reads ...')
+  # Create "profile_id" column.
   sample_meta %<>% tidyr::unite('profile_id', all_of(id_cols), sep=':', remove=F, na.rm=F)
+  # Join cell_set_meta and cell_line_meta. The cell_set can be a name "P939" or a list of LUAs.
   template= sample_meta %>% dplyr::left_join(cell_set_meta, by= 'cell_set') %>%
     dplyr::mutate(members= ifelse(is.na(members), str_split(cell_set, ';'), str_split(members, ';'))) %>% 
-    tidyr::unnest(cols= c(members)) %>%
-    dplyr::left_join(cell_line_meta, by= dplyr::join_by('members'=='LUA'), relationship= 'many-to-one')
+    tidyr::unnest(cols= members) %>%
+    dplyr::left_join(cell_line_meta, by= dplyr::join_by('members'=='LUA'), relationship= 'many-to-one') %>%
+    dplyr::distinct() # to remove duplicate cell lines - cell sets sometime have the same cell line repeated.
```

Comment:
if there are duplicate cell lines in the same cell_set (not project), that could be an issue. (especially if the duplicates were from different pools).
is it possible to print out a warning and the duplicate entries here before doing the distinct operation?

Reply:
Added a validator for cell_set_meta